lib/astutils.cpp

```diff
@@ -1480,8 +1480,16 @@ bool isSameExpression(bool cpp, bool macro, const Token *tok1, const Token *tok2
     if (!tok_str_eq) {
         const Token* refTok1 = followReferences(tok1, errors);
         const Token* refTok2 = followReferences(tok2, errors);
-        if (refTok1 != tok1 || refTok2 != tok2)
+        if (refTok1 != tok1 || refTok2 != tok2) {
+            if (refTok1 && !refTok1->varId() && refTok2 && !refTok2->varId()) { // complex reference expression
+                const Token *start = refTok1, *end = refTok2;
+                if (!precedes(start, end))
+                    std::swap(start, end);
+                if (isExpressionChanged(start, start, end, nullptr, cpp))
```

Comment:
This should probably be done in `followReferences`.

Reply:
What should `start` and `end` then be, since `followReferences()` only takes one token?

Reply:
I assumed `start=nextAfterAstRightmostLeaf(refTok1)` and `end=tok1`, but I dont think that will work well in `followReferences`. This only matters when comparing two expressions are the same, so its probably best to leave it here.

However, there could be a problem. `isExpressionChanged` only works after `createSymbolDatabaseExprIds` is called, but `createSymbolDatabaseExprIds` needs to use `isSameExpression`. So I am not sure the best solution for this. I think we need add another flag but this function needs to be refactored to make it easier to add new flags. Perhaps have different named functions or use a struct to have all the flags similar to `stringifyOptions`.

Reply:
It would also be nice to document what `isSameExpression()` actually does. Expressions can look the same and not be equivalent (as is the case here), but also look different despite being equivalent. So what does 'same' mean?